### PR TITLE
kv: acquire clock reading for lease request under lock

### LINKED ISSUE #134171: kv/kvserver: TestRequestsOnLaggingReplica failed
kv/kvserver.TestRequestsOnLaggingReplica [failed](https://tanzanite.cluster.engflow.com/invocations/default/0e83f18c-77f2-413e-96f7-d8e1e5218580?testReportRun=22&testReportShard=48&testReportAttempt=1#targets-Ly9wa2cva3Yva3ZzZXJ2ZXI6a3ZzZXJ2ZXJfdGVzdA==) on master @ [6bb6dc96ebf0ee2f23c5c568fa0d421019dc0946](https://github.com/cockroachdb/cockroach/commits/6bb6dc96ebf0ee2f23c5c568fa0d421019dc0946):

```
    client_raft_test.go:1308: got [NotLeaseHolderError] refusing to acquire lease on follower; r70: replica (n1,s1):1 not lease holder; lease holder unknown, retrying
    client_raft_test.go:1308: got [NotLeaseHolderError] refusing to acquire lease on follower; r70: replica (n1,s1):1 not lease holder; lease holder unknown, retrying
    client_raft_test.go:1308: got [NotLeaseHolderError] refusing to acquire lease on follower; r70: replica (n1,s1):1 not lease holder; lease holder unknown, retrying
    client_raft_test.go:1308: got [NotLeaseHolderError] refusing to acquire lease on follower; r70: replica (n1,s1):1 not lease holder; lease holder unknown, retrying
    client_raft_test.go:1308: got [NotLeaseHolderError] refusing to acquire lease on follower; r70: replica (n1,s1):1 not lease holder; lease holder unknown, retrying
    client_raft_test.go:1308: got [NotLeaseHolderError] refusing to acquire lease on follower; r70: replica (n1,s1):1 not lease holder; lease holder unknown, retrying
    client_raft_test.go:1308: got [NotLeaseHolderError] refusing to acquire lease on follower; r70: replica (n1,s1):1 not lease holder; lease holder unknown, retrying
    client_raft_test.go:1308: got [NotLeaseHolderError] refusing to acquire lease on follower; r70: replica (n1,s1):1 not lease holder; lease holder unknown, retrying
    client_raft_test.go:1308: got [NotLeaseHolderError] refusing to acquire lease on follower; r70: replica (n1,s1):1 not lease holder; lease holder unknown, retrying
    client_raft_test.go:1308: got [NotLeaseHolderError] refusing to acquire lease on follower; r70: replica (n1,s1):1 not lease holder; lease holder unknown, retrying
    client_raft_test.go:1308: got [NotLeaseHolderError] refusing to acquire lease on follower; r70: replica (n1,s1):1 not lease holder; lease holder unknown, retrying
    client_raft_test.go:1308: got [NotLeaseHolderError] refusing to acquire lease on follower; r70: replica (n1,s1):1 not lease holder; lease holder unknown, retrying
    client_raft_test.go:1308: got [NotLeaseHolderError] refusing to acquire lease on follower; r70: replica (n1,s1):1 not lease holder; lease holder unknown, retrying
    client_raft_test.go:1308: got [NotLeaseHolderError] refusing to acquire lease on follower; r70: replica (n1,s1):1 not lease holder; lease holder unknown, retrying
    client_raft_test.go:1308: got [NotLeaseHolderError] refusing to acquire lease on follower; r70: replica (n1,s1):1 not lease holder; lease holder unknown, retrying
    client_raft_test.go:1308: got [NotLeaseHolderError] refusing to acquire lease on follower; r70: replica (n1,s1):1 not lease holder; lease holder unknown, retrying
    client_raft_test.go:1308: got [NotLeaseHolderError] refusing to acquire lease on follower; r70: replica (n1,s1):1 not lease holder; lease holder unknown, retrying
    client_raft_test.go:1308: got [NotLeaseHolderError] refusing to acquire lease on follower; r70: replica (n1,s1):1 not lease holder; lease holder unknown, retrying
    client_raft_test.go:1308: got [NotLeaseHolderError] refusing to acquire lease on follower; r70: replica (n1,s1):1 not lease holder; lease holder unknown, retrying
    client_raft_test.go:1308: got [NotLeaseHolderError] refusing to acquire lease on follower; r70: replica (n1,s1):1 not lease holder; lease holder unknown, retrying
    client_raft_test.go:1308: got [NotLeaseHolderError] refusing to acquire lease on follower; r70: replica (n1,s1):1 not lease holder; lease holder unknown, retrying
    client_raft_test.go:1308: got [NotLeaseHolderError] refusing to acquire lease on follower; r70: replica (n1,s1):1 not lease holder; lease holder unknown, retrying
    client_raft_test.go:1308: got [NotLeaseHolderError] refusing to acquire lease on follower; r70: replica (n1,s1):1 not lease holder; lease holder unknown, retrying
    client_raft_test.go:1308: got [NotLeaseHolderError] refusing to acquire lease on follower; r70: replica (n1,s1):1 not lease holder; lease holder unknown, retrying
    client_raft_test.go:1308: got [NotLeaseHolderError] refusing to acquire lease on follower; r70: replica (n1,s1):1 not lease holder; lease holder unknown, retrying
    client_raft_test.go:1308: got [NotLeaseHolderError] refusing to acquire lease on follower; r70: replica (n1,s1):1 not lease holder; lease holder unknown, retrying
    client_raft_test.go:1308: got [NotLeaseHolderError] refusing to acquire lease on follower; r70: replica (n1,s1):1 not lease holder; lease holder unknown, retrying
    client_raft_test.go:1308: got [NotLeaseHolderError] refusing to acquire lease on follower; r70: replica (n1,s1):1 not lease holder; lease holder unknown, retrying
    client_raft_test.go:1308: got [NotLeaseHolderError] refusing to acquire lease on follower; r70: replica (n1,s1):1 not lease holder; lease holder unknown, retrying
    client_raft_test.go:1308: got [NotLeaseHolderError] refusing to acquire lease on follower; r70: replica (n1,s1):1 not lease holder; lease holder unknown, retrying
    client_raft_test.go:1308: got [NotLeaseHolderError] refusing to acquire lease on follower; r70: replica (n1,s1):1 not lease holder; lease holder unknown, retrying
    client_raft_test.go:1308: got [NotLeaseHolderError] refusing to acquire lease on follower; r70: replica (n1,s1):1 not lease holder; lease holder unknown, retrying
    client_raft_test.go:1308: got [NotLeaseHolderError] refusing to acquire lease on follower; r70: replica (n1,s1):1 not lease holder; lease holder unknown, retrying
    client_raft_test.go:1308: got [NotLeaseHolderError] refusing to acquire lease on follower; r70: replica (n1,s1):1 not lease holder; lease holder unknown, retrying
    client_raft_test.go:1308: got [NotLeaseHolderError] refusing to acquire lease on follower; r70: replica (n1,s1):1 not lease holder; lease holder unknown, retrying
    client_raft_test.go:1308: got [NotLeaseHolderError] refusing to acquire lease on follower; r70: replica (n1,s1):1 not lease holder; lease holder unknown, retrying
    client_raft_test.go:1308: got [NotLeaseHolderError] refusing to acquire lease on follower; r70: replica (n1,s1):1 not lease holder; lease holder unknown, retrying
    client_raft_test.go:1308: got [NotLeaseHolderError] refusing to acquire lease on follower; r70: replica (n1,s1):1 not lease holder; lease holder unknown, retrying
    client_raft_test.go:1308: got [NotLeaseHolderError] refusing to acquire lease on follower; r70: replica (n1,s1):1 not lease holder; lease holder unknown, retrying
    client_raft_test.go:1308: got [NotLeaseHolderError] refusing to acquire lease on follower; r70: replica (n1,s1):1 not lease holder; lease holder unknown, retrying
    client_raft_test.go:1308: got [NotLeaseHolderError] refusing to acquire lease on follower; r70: replica (n1,s1):1 not lease holder; lease holder unknown, retrying
    client_raft_test.go:1298: 
        	Error Trace:	pkg/kv/kvserver/client_raft_test.go:1298
        	            				pkg/testutils/subtest.go:23
        	Error:      	Should be in error chain:
        	            	expected: %!q(**kvpb.NotLeaseHolderError=0xc00553f730)
        	            	in chain: 
        	Test:       	TestRequestsOnLaggingReplica/symmetric=true
        	Messages:   	lease cannot start after now
    --- FAIL: TestRequestsOnLaggingReplica/symmetric=true (30.69s)
```

Parameters:
 - <code>attempt=1</code>
 - <code>run=22</code>
 - <code>shard=48</code>
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
/cc @cockroachdb/kv
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestRequestsOnLaggingReplica.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/bazci/githubpost/issues)

</sub>

Jira issue: CRDB-43955

### PR DESCRIPTION
**Pull request by nvb**
Fixes #134171.

This commit adds validation that lease construction that the min lease proposed timestamp is never in advance of the current time.

The commit then fixes a case where this could occur by acquiring a clock reading under the replica mutex in redirectOnOrAcquireLeaseForRequest.

While I was never able to reproduce the issue in `TestRequestsOnLaggingReplica`, I did independently and reliably hit the same issue when adding lease type changes to `kvnemesis` (#125260, PR to follow soon) and confirmed that this change fixes the issue.

Epic: None

### CODE REVIEW

**Review by arulajmani — APPROVED** [review]
:lgtm:

> Reviewed 3 of 3 files at r1, 2 of 2 files at r2, 3 of 3 files at r3, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/135042#-:-OBXpl7578gb-igEjCYJ:bnfp4nl)* status: <span>:shipit:</span> complete! 1 of 0 LGTMs obtained (waiting on @nvanbenschoten and @tbg)

___
*[`pkg/kv/kvserver/leases/build.go` line 607 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/135042#-OBXpMc22at13JC4PYG-:-OBXpMc22at13JC4PYG0:bpicndd) ([raw file](https://github.com/cockroachdb/cockroach/blob/bd3bf35853198607329bf88dbabf0df339342581/pkg/kv/kvserver/leases/build.go#L607)):*
> ```Go
> 		prevType := i.PrevLease.Type()
> 		indirectExp := prevType != roachpb.LeaseExpiration
> 		switchingType := prevType != nextLease.Type()
> ```

Were you able to figure out why `switchingType` was true? Were we acquiring an expiration based lease for some reason?

**Review by nvb — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/135042)* status: <span>:shipit:</span> complete! 1 of 0 LGTMs obtained (waiting on @arulajmani and @tbg)

___
*[`pkg/kv/kvserver/leases/build.go` line 607 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/135042#-OBXpMc22at13JC4PYG-:-OBXpxhC3mvoQgfJbjf_:b-s665i1) ([raw file](https://github.com/cockroachdb/cockroach/blob/bd3bf35853198607329bf88dbabf0df339342581/pkg/kv/kvserver/leases/build.go#L607)):*
<details><summary><i>Previously, arulajmani (Arul Ajmani) wrote…</i></summary><blockquote>

Were you able to figure out why `switchingType` was true? Were we acquiring an expiration based lease for some reason?
</blockquote></details>

What I think I saw happen was that a leader lease expired when a raft leadership term ended. The replica then attempted to acquire the lease again, and hit this path because it was a follower: https://github.com/cockroachdb/cockroach/blob/2520c51663e57de54706a58e8e68370467dbe871/pkg/kv/kvserver/leases/build.go#L387

So once the leader lease expired, it tried to acquire an expiration based lease. I think we will allow this until we get `kv.lease.reject_on_leader_unknown.enabled` defaulted to true (https://github.com/cockroachdb/cockroach/issues/118435).

**Review by arulajmani — APPROVED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/135042)* status: <span>:shipit:</span> complete! 1 of 0 LGTMs obtained (waiting on @nvanbenschoten and @tbg)

___
*[`pkg/kv/kvserver/leases/build.go` line 607 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/135042#-OBXpMc22at13JC4PYG-:-OBXtZwx9GWxUyg76J_8:bwuv9up) ([raw file](https://github.com/cockroachdb/cockroach/blob/bd3bf35853198607329bf88dbabf0df339342581/pkg/kv/kvserver/leases/build.go#L607)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

What I think I saw happen was that a leader lease expired when a raft leadership term ended. The replica then attempted to acquire the lease again, and hit this path because it was a follower: https://github.com/cockroachdb/cockroach/blob/2520c51663e57de54706a58e8e68370467dbe871/pkg/kv/kvserver/leases/build.go#L387

So once the leader lease expired, it tried to acquire an expiration based lease. I think we will allow this until we get `kv.lease.reject_on_leader_unknown.enabled` defaulted to true (https://github.com/cockroachdb/cockroach/issues/118435).
</blockquote></details>

Nice!

File: pkg/kv/kvserver/replica_range_lease.go [hunk 1]
```diff
@@ -1245,7 +1245,6 @@ func (r *Replica) redirectOnOrAcquireLeaseForRequest(
 	// Loop until the lease is held or the replica ascertains the actual lease
 	// holder. Returns also on context.Done() (timeout or cancellation).
 	for attempt := 1; ; attempt++ {
-		now = r.store.Clock().NowAsClockTimestamp()
 		llHandle, status, transfer, pErr := func() (*leaseRequestHandle, kvserverpb.LeaseStatus, bool, *kvpb.Error) {
```
Comment (tbg) [inline pkg/kv/kvserver/replica_range_lease.go:1248]: so what was the (or at least a possible) sequence of events here?

1. we grab `now` above
2. someone bumps the minLeaseProposedTS. Is this someone someone else calling `redirectOnOrAcquireLeaseForRequest` and racing ahead of us, which (during eval) hits `RevokeLease`[^1]?
3. but then we would join the ongoing request already, no? And that would prevent us from calling `VerifyAndBuild`, so we wouldn't error out?

[^1]: https://github.com/cockroachdb/cockroach/blob/90d1c8caa24033e6a2f77c768dab0378fa31f514/pkg/kv/kvserver/replica_range_lease.go#L1030-L1036

**Review by tbg — APPROVED** [review]

File: pkg/kv/kvserver/replica_range_lease.go (hunk 1 above)
Comment (nvanbenschoten) [reply pkg/kv/kvserver/replica_range_lease.go:1248]: > Is this someone someone else calling redirectOnOrAcquireLeaseForRequest and racing ahead of us, which (during eval) hits RevokeLease[1](https://github.com/cockroachdb/cockroach/pull/135042#user-content-fn-1-53155447d858ae7b5fdcf644d14e8687)?

Yes, I added some additional logging to kvnemesis and confirmed that it is.

> but then we would join the ongoing request already, no? And that would prevent us from calling VerifyAndBuild, so we wouldn't error out?

It looks like the ongoing request completes (unsuccessfully) before the next request acquires the replica mutex. `request lease: err=[NotLeaseHolderError] ‹refusing to acquire lease on follower›`. So there's nothing for it to join to.

This timeline is strange, but there's nothing wrong with it besides the fact that we are using a stale clock timestamp for `now`, which this commit fixes.

**Comment by nvanbenschoten** [comment]
TFTR!

bors r+

**Comment by nvanbenschoten** [comment]
blathers backport v24.3.0

**Comment by nvanbenschoten** [comment]
blathers backport 24.3.0-rc

**Pull request merged**